RationaleMCP/0036/README.md

```diff
@@ -39,19 +53,46 @@ One goal of "Clocked Discretized Continuous-Time Partition" is to preserve the b
 In addition the current reinit might possibly in the future be replaced by some form of impulses, indicating that the changes are physical and influence other parts of the system, this does not seem consistent with how states are set in this MCP.
 Finally having a conditional setting of states is more complicated, and reinit and other similar operators naturally suggest that - and we either need to determine what it means or forbid it.
 
+## Proposed alternative
+Use a special binding equation:
+
+Real x=measurement(xs);
+
+This special operator, measurement, is only legal in a "Clocked Discretized Continuous-Time Partition" and only in this way and means that 'x' is first seen as a state, but after index reduction x is no longer a state and x=xs replaces the integration of der(x).
```

Comment:
This part is what concerns me most.  Phrases like _first seen as a state_ and references to index reduction makes me wonder if there is a solid mathematical foundation to rest upon here.  What mathematical truths will still hold, and which may no longer hold in the presence of implementation-oriented specification like this?

While `reinit` also has its share of loose mathematical grounds, perhaps a better approach here could be to build upon `reinit` rather than introducing one more source of mathematical inconsistency.  That is, one wouldn't replace the integration of `der(x)`, but rather make sure that the integrated value can be overridden with a different value at any (every) clock tick.  This way, `der(x)` can still be the derivative of `x` between clock ticks, but at the end of a clock evaluation, `x` may jump in a way similar to how `reinit` can cause a variable to jump at the end of an event iteration.

If we do it right, I still believe that it would be possible for a tool to see that the values for `x` computed by integration of `der(x)` might never be used, and hence drop the useless computations to get the same performance as with the current proposal.

Reply:
For getting a physical understanding of reinit the problem the problem I see is that there seems to be several different uses of reinit, and trying to combine them is non-trivial. And now we have yet another variant.

- Modelica.Mechanics.Translational.Components.MassWithStopAndFriction: simplified modeling of a physical impulse. This impulse could be transferred through the normal connectors, but the problem is that impulse handling is in general not well-defined. (I believe there are some old papers about this.)
- Modelica.Blocks.Math.Mean: a mathematical trick to improve accuracy. In this case we want to take the difference between the integral at t and t-h, and setting the integral to zero avoids accumulating errors - but is not needed in a strict mathematical sense; and there is thus no physical thing to understand.
- Modelica.Magnetic.FluxTubes.Shapes.HysteresisAndMagnets.GenericHystPreisachEverett: for hysteresis in some way in magnets, how do they work?

Technically speaking Mass... has two reinit, one that would correspond to the impulse (for v) and one that is more of a trick to improve accuracy (for s) and which strictly speaking shouldn't be needed.

Reply:
> This part is what concerns me most. Phrases like _first seen as a state_ and references to index reduction makes me wonder if there is a solid mathematical foundation to rest upon here. What mathematical truths will still hold, and which may no longer hold in the presence of implementation-oriented specification like this?

To me the explanation for this is that it is treated as normal up to the discretization based on the solver method:
https://specification.modelica.org/master/synchronous-language-elements.html#solver-methods
And instead of using one of these discretization formulas the state is equal to the measurement-value during the entire step. (The same should also work for "External" - but there might be some caveats.)

For this to make sense:

- It must be a continuous variable in a clocked partition.
- It must be a state.

If this procedure isn't well-defined it seems the problem is in the discretization based on solver-method, and that is already present in the specification.

Reply:
OK, I think that makes sense.